### PR TITLE
curl: make --libcurl show binary posts correctly

### LINKED ISSUE #6031: --libcurl fails to reproduce binary data from --data-binary input that contains null bytes
### I did this

I came across this when I tried to convert a curl command line tool invocation into a corresponding libcurl program. The use case was to retrieve a time-stamp from a time-stamp server. The attached file  is only for demonstration purposes and does not contain an actual time-stamp request. The file contains null bytes and is six bytes large:

```
E:\Users\stm\Documents\GitHub\curl [master ≡ +4 ~0 -0 !]> wsl xxd bindata.gif
00000000: 6100 0000 0062
```

curl command line tool invocation:

```
E:\Users\stm\Documents\GitHub\curl [master ≡ +4 ~0 -0 !]> .\build\src\Debug\curl.exe --libcurl - -s -H "Content-Type: application/timestamp-query" --data-binary "@bindata.gif" -o response.tsr http://zeitstempel.dfn.de
/********* Sample code generated by the curl command line tool **********
 * All curl_easy_setopt() options are documented at:
 * https://curl.haxx.se/libcurl/c/curl_easy_setopt.html
 ************************************************************************/
...
  curl_easy_setopt(hnd, CURLOPT_POSTFIELDS, "a");
  curl_easy_setopt(hnd, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)6);
...
```

Note that the value supplied for `CURLOPT_POSTFIELDSIZE_LARGE` does not match the length of the data supplied via `CURLOPT_POSTFIELDS`.

### I expected the following

I expected that the string passed via `CURLOPT_POSTFIELDS` contains the full data from the input file, e.g.:

```
curl_easy_setopt(hnd, CURLOPT_POSTFIELDS, "a\000\000\000\000b");
```
### curl/libcurl version

Curl built from commit 7920be94738e0d84f7aa30c3605d29ec4274c714.

```
E:\Users\stm\Documents\GitHub\curl [master ≡ +4 ~0 -0 !]> .\build\src\Debug\curl.exe -V
curl 7.73.0-DEV (Windows) libcurl/7.73.0-DEV
Release-Date: [unreleased]
Protocols: dict file ftp gopher http imap ldap mqtt pop3 rtsp smb smtp telnet tftp
Features: AsynchDNS IPv6 Largefile NTLM
```

### operating system

Windows 10 Version 2004, but also reproduced on Linux.

### PR DESCRIPTION
**Pull request by bagder**
Reported-by: Stephan Mühlstrasser
Fixes #6031

### CODE REVIEW

File: src/tool_setopt.c [hunk 1]
```diff
@@ -219,26 +219,34 @@ static const struct NameValue setopt_nv_CURLNONZERODEFAULTS[] = {
...
+
+#define MAX_STRING_LENGTH_OUTPUT 2000
+#define ZERO_TERMINATED -1
+
+static char *c_escape(const char *str, curl_off_t len)
 {
   const char *s;
   unsigned char c;
   char *escaped, *e;
+  int cutoff = 0;
```
Comment (emilengler) [inline src/tool_setopt.c:231]: I would make this a uint

File: src/tool_setopt.c [hunk 2]
```diff
@@ -219,26 +219,34 @@ static const struct NameValue setopt_nv_CURLNONZERODEFAULTS[] = {
...
-    return NULL;
+  if(len > MAX_STRING_LENGTH_OUTPUT) {
+    /* cap ridiculously long strings */
+    len = MAX_STRING_LENGTH_OUTPUT;
+    cutoff = 3;
+  }
 
   /* Allocate space based on worst-case */
-  escaped = malloc(4 * len + 1);
+  escaped = malloc(4 * (size_t)len + 1 + cutoff);
```
Comment (emilengler) [inline src/tool_setopt.c:243]: Keep math in mind. I'm not 100% sure what you are doing but I think it's more secure to put the addition into brackets.

Comment (bagder) [reply src/tool_setopt.c:243]: As you can see it wasn't within any parentheses before. The math expression counts the largest possible area needed for an escaped string and I believe it is fine as is.

File: src/tool_setopt.c (hunk 1 above)
Comment (bagder) [reply src/tool_setopt.c]: Agreed!

File: src/tool_setopt.c (hunk 2 above)
Comment (emilengler) [reply src/tool_setopt.c:243]: Yeah I am not into that part of the code but it is a typical mistake often made ;)

**Pull request closed without merging**